### ISSUE TITLE
Cognito calls are very slow in ios

### ISSUE THREAD

**Issue by JackLewisGit** [ionic-team/ionic2-starter-aws#5]
This is a problem my team has experienced for some time (before this repo was created). From what I understand the encryption done by the cognito/aws (not sure which one) sdk doesn't play nice with the JavaScriptCore. Executing a login or register will hang for up to 15-20 seconds on older ios devices.

I get that it's not an ionic problem but I'd be interested in hearing workarounds or ideas from others.

**Comment by mlynch** [comment]
Thanks, we're looking into using the native aws sdk for this routine, we heard this from other projects using JS SDK

**Comment by lukezbikowski** [comment]
If you go with WKWebView it's much better. There is still a little 'notch' thou... 
See 
https://github.com/Hawatel/aws-cognito-ionic-demo/issues/1
https://github.com/aws/amazon-cognito-identity-js/issues/209
https://github.com/aws/amazon-cognito-identity-js/issues/69
https://github.com/aws/amazon-cognito-identity-js/issues/179
for reference. 
Another option is using ADMIN_NO_SRP_AUTH (what is not the best approach imho).

**Comment by jeffleus** [comment]
I have had similar experiences and tried to write a plugin to process the Cognito stuff natively as a plugin.  I started by writing a native app using the AWS native SDK in objective-c.  The basic features worked: init, login, refresh and logout.  And, the refresh specifically works by returning the current session until the expiration is reached and then a new session is returned with a new idToken and expiration.

But as luck has it, when I wrap the code as a plugin, things fail.  The init and login still work correctly.  But, something is failing during the refresh.  My guess is that some internal state is being lost between the call to init/login and the call to refresh.  @mlynch I am wondering if you have any updates on your progress w/ the native processing.

Were you also planning to take a plugin approach?  Do you have any updates that you can share?  I have a project that was built with a full AWS back-end stack (RDS data, lambda data access, API gateway with Cognito authorization built in).  But it fails miserably on an actual iOS device as the javascript sdk takes 60sec to complete logins and refreshing the tokens.  Uugghh!  Any help, clues, links, advice, etc would be greatly appreciated.

**Comment by F1LT3R** [comment]
@jeffleus we think we are experiencing the same thing in our project. It looks like @jmparsons hit this issue in a React-Native project and replaced the Big Int math on iOS, because the BitInt provided by AWS was too slow. https://github.com/AirLabsTeam/react-native-aws-cognito-js

**Comment by F1LT3R** [comment]
@jeffleus https://github.com/aws/amazon-cognito-identity-js/issues/69

**Comment by jeffleus** [comment]
@F1LT3R I will take a look at the work by @jmparsons, but I am not a react guy so the native bridge stuff may be tough to follow.  in fact, I am not much of a native obj-c guy but was able to get up and running w/ the AWS library.  My current "fix" is a super kludgey work-around where I cache the username/pwd credentials and re-login before expiration.  It runs fast when executed w/ the native iOS libraries from AWS.  BUT... it is driving me crazy that the work-around is needed.

Do you have any objective-c/native skills?  Any interest in helping me tackle this?  I am really close and all my code works as a full native app.  But when I wrap the code to run as a plugin, something is breaking.  There is some state in the iOS/AWS library that is getting lost between the login and the refresh piece...

**Comment by F1LT3R** [comment]
We are currently looking into using WKWebView with Ionic instead of React, more on that here: https://github.com/Hawatel/aws-cognito-ionic-demo/issues/1 - if we have any success with that I will let you know.

**Comment by F1LT3R** [comment]
@jeffleus - we are currently looking into using WKWebView with Ionic instead of React, more on that here: https://github.com/Hawatel/aws-cognito-ionic-demo/issues/1 - if we have any success with that I will let you know.

**Comment by F1LT3R** [comment]
> Do you have any objective-c/native skills? 

@jeffleus - none, but I'd be interested to try and make it work anyway.

**Comment by jeffleus** [comment]
@F1LT3R Yes, I looked at the WKWebView solution.  The problem that I have is WKWebView caused a number of other problems for me with my existing project and seems like a rathole solution.  The plugin approach should work but something breaks when I wrap the native code as a plugin.

My code is all based off of some AWS resources that I don't want to share.  But, I could re-write as a demo app and share via github.  I've tried to get help via forums and Ionic.  I really thought I would hear something from the Ionic team (where are you @mlynch ?) since they wrote an AWS demo app and partnered w/ AWS.  They have the same issue identified and mentioned looking into it.  And... I built my plugin off of @mlynch article on how to write cordova plugins.  But I haven't heard anything back from them.  I even spoke w/ an AWS support engineer. But he basically threw up his hands and said my problem was w/ cordova.  I am really close to completing a solution that could be used as a cordova plugin.  The login works and the tokens can be collected and used in the http headers for calls to my serverless api.  But the refresh token process is where things fail.

If I setup the demo, do you have some time to help diagnose and brainstorm?

**Comment by mlynch** [comment]
I'm following this issue :) We're working on a number of other things right now and will get back to this once we have some resources freed up. We are continuing to work with AWS and there are some other things in motion that should be interesting to people using both services. Stay tuned

**Comment by peterecofit** [comment]
Any news?  We're using AWS Cognito and Ionic 2, everything is great aside from this!

**Comment by mlynch** [comment]
Has anyone tried this with Our new WKWebView plugin? Updated a few weeks ago: https://www.npmjs.com/package/cordova-plugin-ionic-webview

**Comment by lukezbikowski** [comment]
@jeffleus I don't have any significant issues with WKWebView (anymore). If there is any particular issue - imho it's worth raising here... :-) My view - go with WkWebView, fix issues/find workarrounds and don't look back..

Also - I was considering writing native plugin but it's wrong idea. Number of changes in AWS is significant (see the AWS Blog :-) ) ... I don't believe you want to maintain it and keep inline with changes... 

I use quite a few services + call Serverless API directly... no significant issues so far.

**Comment by le0366** [comment]
@mlynch I tried this and it fixed my problem in AWS Cognito.  However, I encountered an issue with my SELECT statement using SQLITE.  I've posted a topic in ionic forum for this problem [https://forum.ionicframework.com/t/select-statement-not-working-after-installing-wkwebview/109660](url).

**Comment by F1LT3R** [comment]
As a follow-up, we found that adding WKWebView to the iOS Cordova project reduced our login time from `~50` seconds to `~5` seconds on an iPhone 5 with iOS 10.3.3.

We are using the Cordova plugin `cordova-plugin-wkwebview-engine` with `cordova-ios` platform `4.5.4` which is the current version when using `cordova platform add ios@latest`.

Note: the following code must be added to the `config.xml` to make WKWebView work correctly:

```xml
<feature name="CDVWKWebViewEngine">
    <param name="ios-package" value="CDVWKWebViewEngine" />
</feature>
<preference name="CordovaWebViewEngine" value="CDVWKWebViewEngine" />
```

We have found that about 5 seconds was an acceptable login time.

I hope that helps someone.
